File: templates/ObjectiveC-UIKit/Features/Settings/SettingsViewController.m
//
//  SettingsViewController.m
//  ObjectiveC-UIKit
//

#import "SettingsViewController.h"
#import "UIView+Extensions.h"

// NSUserDefaults keys backing each toggle. Keep these stable across releases.
static NSString *const kDefaultsNotificationsEnabled = @"settings.notificationsEnabled";
static NSString *const kDefaultsHapticsEnabled = @"settings.hapticsEnabled";
static NSString *const kDefaultsAnalyticsEnabled = @"settings.analyticsEnabled";

static NSString *const kCellReuseIdentifier = @"SettingsToggleCell";

@interface SettingsViewController () <UITableViewDataSource>

@property (nonatomic, strong) UITableView *tableView;

// Each row is described by a title + the defaults key it toggles.
@property (nonatomic, copy) NSArray<NSDictionary<NSString *, NSString *> *> *rows;

@end

@implementation SettingsViewController

#pragma mark - Lifecycle

- (instancetype)init {
    self = [super initWithNibName:nil bundle:nil];
    if (self) {
        _rows = @[
            @{ @"title": @"Notifications", @"key": kDefaultsNotificationsEnabled },
            @{ @"title": @"Haptic Feedback", @"key": kDefaultsHapticsEnabled },
            @{ @"title": @"Share Analytics", @"key": kDefaultsAnalyticsEnabled },
        ];
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];

    self.title = @"Settings";
    self.view.backgroundColor = [UIColor systemGroupedBackgroundColor];

    UITableView *tableView =
        [[UITableView alloc] initWithFrame:CGRectZero
                                     style:UITableViewStyleInsetGrouped];
    tableView.dataSource = self;
    [tableView registerClass:[UITableViewCell class]
      forCellReuseIdentifier:kCellReuseIdentifier];
    [self.view addSubview:tableView];
    [tableView pinEdgesToView:self.view];
    self.tableView = tableView;
}

#pragma mark - Toggle Handling

- (void)switchValueChanged:(UISwitch *)sender {
    NSDictionary<NSString *, NSString *> *row = self.rows[(NSUInteger)sender.tag];
    NSString *key = row[@"key"];
    [[NSUserDefaults standardUserDefaults] setBool:sender.isOn forKey:key];
    // TODO: React to specific setting changes (register for notifications,
    // toggle analytics SDK, etc.).
}

#pragma mark - UITableViewDataSource

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView
 numberOfRowsInSection:(NSInteger)section {
    return (NSInteger)self.rows.count;
}

- (nullable NSString *)tableView:(UITableView *)tableView
         titleForFooterInSection:(NSInteger)section {
    return @"Preferences are stored on this device.";
}

- (UITableViewCell *)tableView:(UITableView *)tableView
         cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell =
        [tableView dequeueReusableCellWithIdentifier:kCellReuseIdentifier
                                        forIndexPath:indexPath];

    NSDictionary<NSString *, NSString *> *row = self.rows[(NSUInteger)indexPath.row];

    UIListContentConfiguration *content = [cell defaultContentConfiguration];
    content.text = row[@"title"];
    cell.contentConfiguration = content;
    cell.selectionStyle = UITableViewCellSelectionStyleNone;

    UISwitch *toggle = [[UISwitch alloc] init];
    toggle.tag = indexPath.row;
    toggle.on = [[NSUserDefaults standardUserDefaults] boolForKey:row[@"key"]];
    [toggle addTarget:self
               action:@selector(switchValueChanged:)
     forControlEvents:UIControlEventValueChanged];
    cell.accessoryView = toggle;

    return cell;
}

@end

File: templates/Swift-UIKit/Features/Settings/SettingsViewController.swift
import UIKit
import Observation

/// Programmatic UIKit settings screen with toggles backed by the view model.
@MainActor
final class SettingsViewController: UIViewController {

    private let viewModel: SettingsViewModel
    private let tableView = UITableView(frame: .zero, style: .insetGrouped)

    private enum Row: Int, CaseIterable {
        case notifications
        case darkMode
        case version
    }

    init(viewModel: SettingsViewModel) {
        self.viewModel = viewModel
        super.init(nibName: nil, bundle: nil)
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        title = "Settings"
        view.backgroundColor = .systemBackground

        tableView.dataSource = self
        tableView.register(UITableViewCell.self, forCellReuseIdentifier: "Cell")
        view.addSubview(tableView)
        tableView.translatesAutoresizingMaskIntoConstraints = false
        NSLayoutConstraint.activate([
            tableView.topAnchor.constraint(equalTo: view.topAnchor),
            tableView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
            tableView.trailingAnchor.constraint(equalTo: view.trailingAnchor),
            tableView.bottomAnchor.constraint(equalTo: view.bottomAnchor),
        ])
    }

    // MARK: - Toggle handling

    @objc private func notificationsToggled(_ sender: UISwitch) {
        viewModel.isNotificationsEnabled = sender.isOn
    }

    @objc private func darkModeToggled(_ sender: UISwitch) {
        viewModel.isDarkModePreferred = sender.isOn
        // TODO: Apply the user's appearance preference to the active window scene.
    }

    private func makeSwitch(isOn: Bool, action: Selector) -> UISwitch {
        let toggle = UISwitch()
        toggle.isOn = isOn
        toggle.addTarget(self, action: action, for: .valueChanged)
        return toggle
    }
}

// MARK: - UITableViewDataSource

extension SettingsViewController: UITableViewDataSource {

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        Row.allCases.count
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "Cell", for: indexPath)
        var config = cell.defaultContentConfiguration()
        cell.selectionStyle = .none
        cell.accessoryView = nil

        switch Row(rawValue: indexPath.row) {
        case .notifications:
            config.text = "Notifications"
            cell.accessoryView = makeSwitch(
                isOn: viewModel.isNotificationsEnabled,
                action: #selector(notificationsToggled(_:))
            )
        case .darkMode:
            config.text = "Prefer Dark Mode"
            cell.accessoryView = makeSwitch(
                isOn: viewModel.isDarkModePreferred,
                action: #selector(darkModeToggled(_:))
            )
        case .version:
            config.text = "Version"
            config.secondaryText = viewModel.appVersion
        case .none:
            break
        }

        cell.contentConfiguration = config
        return cell
    }
}
